### PR TITLE
Yacht Dice - limit number of categories to prevent OverflowError on gen with start_inventory etc

### PR DESCRIPTION
**Pull request by spinerak**
Yacht Dice can get categories several times, every time you get one, the score of it is doubled.
When you get 27+ of one category, the score (50 * 2^27) becomes too large for an int and you get an `OverflowError`.
This won't happen when you just follow the yaml settings, but it is possible with specific start_inventory or plando settings etc.

## What is this fixing or adding?
This fixes the OverflowError during generation, if start_inventory or plando settings creates many of the same category.
In logic calculation, we limit the number to 20, which is already way more than necessary in any scenario.

## How was this tested?
There is no error anymore when you start with 100+ categories. Also did several default yaml test generations.

### CODE REVIEW

**Comment by NewSoupVi** [comment]
OverflowError? I thought base python ints are supposed to be able to get arbitrarily large
Weird

**Comment by Berserker66** [comment]
There is no int overflow, comment says max, code says min.

**Comment by spinerak** [comment]
Ah wait I had another option open, I'll double check at what point I get this error...

**Comment by spinerak** [comment]
It seems to happen with 1050+ categories, so I assume 2^1024 is a number that's too big?
The point stands that there's no reason someone would want more than 20 of one category, which for the logic calculation, limiting it to 20 is the safe way to go.

**Comment by NewSoupVi** [comment]
Oh yeah the PR is fine, I was just confused lol

**Review by Exempt-Medic — APPROVED** [review]
The change LGTM (tested a generation and it succeeded). It might be a little confusing that you just don't get them from your starting inventory, so it may be good to print a warning somewhere like `generate_early` if more than 20 are in the start inventory and then actually *change* the value to 20. But that's all up to you

**Comment by NewSoupVi** [comment]
> The change LGTM (tested a generation and it succeeded). It might be a little confusing that you just don't get them from your starting inventory, so it may be good to print a warning somewhere like `generate_early` if more than 20 are in the start inventory and then actually _change_ the value to 20. But that's all up to you

+1 on printing a warning and actually removing the extra copies, it's not necessary but I think it's cleaner personally

**Comment by spinerak** [comment]
A warning makes sense! Although wouldn't it make more sense to add a notification to the game, so that the player sees it instead of the host (since it's not game-breaking or dangerous, just a goofy choice of the player)? 
- I can give a one-time-per-game notification when a player receives their 21st copy of a category.
- And I can give a one-time-per-game notification when a player receives a category that is not active in their game (there are 32 categories in total but only 16 selected ones are active in a game. Receiving one that is not active (can happen because of start inventory / plando) does nothing for logic and nothing in game).

**Comment by Exempt-Medic** [comment]
> It might be a little confusing that you just don't get them from your starting inventory, so it may be good to print a warning somewhere like `generate_early` if more than

Those make sense to me, but I'd recommend having a warning in both `generate_early` and when playing, if the player *wanted* those categories (they put them into `start_inventory` after all), they should be able to be informed that they won't be getting them before the game actually starts.

File: worlds/yachtdice/__init__.py [hunk 1]
```diff
@@ -157,6 +158,19 @@ def generate_early(self):
             else:
                 self.itempool.append(cats[categorylist[index]])
 
+        # Give warnings and change start_inventory if necessary.
+        gave_warning = False
+        for item_name, count in self.multiworld.worlds[self.player].options.start_inventory.value.items():
+            if item_name.startswith("Category") and count > 20:
```
Comment (Exempt-Medic) [inline worlds/yachtdice/__init__.py:164]: Should be able to just do this:
```suggestion
        for item_name, count in self.options.start_inventory.value.items():
            if item_name.startswith("Category") and count > 20:
```

**Comment by spinerak** [comment]
I've added the warnings to the apworld (thanks for the comment!).
I think it's good this way, just one small remark:
- if a player currently wants to start with one of each category, they will have to put all 32 categories in the start_inventory and there will be no way around getting the warning that some categories are not one of the 16 active ones.

I don't this is a deal-breaker and it's probably better to show the warning anyway.

**Comment by spinerak** [comment]
I am closing this pull request because I might change/improve code that relates to this. Merging this piece of code now or in the near future may break more than that it would fix.
Thanks for the comments on this PR!

**Pull request closed without merging**